dipy/reconst/ivim.py

```diff
@@ -0,0 +1,383 @@
+        split_b : split the data at this b-value for a two stage fit
+
+        References
+        ----------
+        .. [1] Le Bihan, Denis, et al. "Separation of diffusion
+                   and perfusion in intravoxel incoherent motion MR
+                   imaging." Radiology 168.2 (1988): 497-505.
+        .. [2] Federau, Christian, et al. "Quantitative measurement
+                   of brain perfusion with intravoxel incoherent motion
+                   MR imaging." Radiology 265.3 (2012): 874-881.
+        """
+        ReconstModel.__init__(self, gtab)
+        self.split_b = split_b
+        self.min_signal = min_signal
+        if self.min_signal is not None and self.min_signal <= 0:
+            e_s = "The `min_signal` key-word argument needs to be strictly"
+            e_s += " positive."
+            raise ValueError(e_s)
+
+    def fit(self, data, mask=None, x0=None, fit_method="one_stage", routine="minimize",
+            jac=None, bounds=((0, None), (0, 1.), (0, 1.), (0, 1.)), tol=1e-25, algorithm='L-BFGS-B',
+            gtol=1e-25, ftol=1e-25, eps=1e-15, maxiter=1000):
+        """ Fit method of the Ivim model class
+
+        Parameters
+        ----------
+        data : array
+            The measured signal from one voxel.
+        mask : array
+            A boolean array used to mark the coordinates in the data that
+            should be analyzed that has the shape data.shape[:-1]
+        x0 : array
+            Initial guesses for the parameters S0, f, D_star and D
+            Default x0 = [1.0, 0.10, 0.001, 0.0009]
+            These guess parameters are taken from the Federau paper
+            Dimension can either be 1 or (N, 4) where N is the number of
+            voxels in the data.
+        fit_method: str
+            Use one-stage fitting or two-stage fitting.
+        jac : Boolean
+            Use the Jacobian
+        bounds : tuple
+            Bounds for the various parameters
+        tol : float
+            tolerance for convergence
+        method : str
+            Fitting algorithm to use from scipy.optimize
+
+        Returns
+        -------
+        IvimFit object
+        """
+        if mask is None:
+            # Flatten it to 2D either way:
+            data_in_mask = np.reshape(data, (-1, data.shape[-1]))
+        else:
+            # Check for valid shape of the mask
+            if mask.shape != data.shape[:-1]:
+                raise ValueError("Mask is not the same shape as data.")
+            mask = np.array(mask, dtype=bool, copy=False)
+            data_in_mask = np.reshape(data[mask], (-1, data.shape[-1]))
+
+        if self.min_signal is None:
+            min_signal = _min_positive_signal(data)
+        else:
+            min_signal = self.min_signal
+
+        # Generate guess_parameters for all voxels
+        if x0 is None:
+            x0 = np.ones(
+                (data.shape[:-1] + (4,))) * [1.0, 0.10, 0.001, 0.0009]
+        else:
```

Comment:
Can we enable the mode in which the user wants to send in a series of guess parameters that are the same for all voxels? Here, you require that a guess be sent for *each* voxel. I'd think that users would usually want to make the same guess everywhere.